File: csc454hw03/src/main/java/NetworkModel.java
import java.util.HashMap;
import java.util.ArrayList;

public class NetworkModel extends Model {

    HashMap<String, Model> children;
    ArrayList<Pipe> pipes;
    int numberOfTicks;

    public NetworkModel(int[] initial, Model inputModel, Model outputModel, int numberOfTicks) {
        super(initial);
        children = new HashMap<String, Model>();
        pipes = new ArrayList<Pipe>();
        this.numberOfInputs = inputModel.numberOfInputs;
        this.inPorts = new Port[numberOfInputs];
        for (int i = 0; i < numberOfInputs; i++) {
            this.inPorts[i] = inputModel.inPorts[i];
        }
        this.outPort = outputModel.outPort;
        this.numberOfTicks = numberOfTicks;

    }

    public void tick(int[] netIn) {
        for (int tick = 0; tick < numberOfTicks; tick++) {
            /** Lambda's */
            lambda();

            /** Pass the value into the networks input Ports */
            for (int i = 0; i < netIn.length; i++) {
                this.inPorts[i].currentValue = netIn[i];
            }

            /** Pass all pipe values around */
            executePipes();

            /** Delta's */
            delta(netIn);
        }

    }

    public int lambda() {
        for (String model : children.keySet()) {
            Model currentModel = children.get(model);
            currentModel.outPort.currentValue = currentModel.lambda();
            debugPrint("Lambda from " + model + ": " + currentModel.lambda());
        }
        return 0;
    }

    public int[] delta(int[] inputSet) {
        for (String model : children.keySet()) {
            Model currentModel = children.get(model);
            int[] inputs = currentModel.convertInPortsToIntArr();
            currentModel.state = currentModel.delta(inputs);
            if (verbose) {
                String state = "";
                for (int i = 0; i < currentModel.state.length; i++) {
                    state += currentModel.state[i] + " ";
                }
                state = state.substring(0, state.length() - 1);
                System.out.println("New state for " + model + ": {" + state + "}");
            }
        }

        return null;
    }

    public void executePipes() {
        for (Pipe p : pipes) {
            p.passValue();
        }
    }

    public void addModel(String modelName, Model m) {
        this.children.put(modelName, m);
    }

    public void addPipe(Pipe p) {
        pipes.add(p);
    }

}

File: csc454hw04_cpp/NetworkModel.cpp
#ifndef NETWORK_MODEL
#define NETWORK_MODEL

#include <map>
#include "Model.cpp"
#include "Pipe.cpp"
#include <string>
#include <vector>
#include<iostream> 

using namespace std;

class NetworkModel : public Model
{

public:
    map<string, Model *> *children;
    map<string, Pipe *> *pipes;
    int numberOfTicks;

    NetworkModel(Model *inputModel, Model *outputModel, int numTicks)
    {
        numberOfInputs = inputModel->numberOfInputs;
        inPorts = new Port*[numberOfInputs];
        for (int i = 0; i < numberOfInputs; i++)
        {
            inPorts[i] = inputModel->inPorts[i];
        }
        outPort = outputModel->outPort;
        numberOfTicks = numTicks;
        pipes = new map<string, Pipe*>();
        children = new map<string, Model*>();
    }

    void tick(int *netIn)
    {
        for (int tick = 0; tick < numberOfTicks; tick++)
        {

            lambda();

            for (int i = 0; i < numberOfInputs; i++)
            {
                inPorts[i]->currentValue = netIn[i];
            }

            executePipes();

            delta(netIn);
        }
    }

    int lambda()
    {
        map<string, Model *>::iterator itr;
        for (itr = children->begin(); itr != children->end(); itr++)
        {
            Model *currentModel = itr->second;
            int value = currentModel->lambda();
            currentModel->outPort->setValue(value);
            debugPrint("Lambda from " + itr->first + ": " + to_string(value));
        }
        return 0;
    }

    void delta(int *inputSet)
    {
        map<string, Model *>::iterator itr;
        for (itr = children->begin(); itr != children->end(); itr++)
        {
            Model * currentModel = itr->second;
            int * inputs = currentModel->convertInPortsToIntArr();
            currentModel->delta(inputs);
            delete inputs;
            if (verbose) {
                string state = "";
                for (int i=0; i< currentModel->stateSize; i++) {
                    state += to_string(currentModel->state[0]) + " ";
                }
                state.pop_back();
                printf("New state for %s : { %s }\n", itr->first.c_str(), state.c_str());
            }
        }
    }

    void executePipes()
    {
        map<string, Pipe *>::iterator itr;
        for (itr = pipes->begin(); itr != pipes->end(); itr++)
        {
            itr->second->passValue();
        }
    }

    void addModel(string modelName, Model *m)
    {
        children->insert(pair<string, Model*>(modelName, m));
    }

    void addPipe(string pipeName, Pipe * pipe) {
        pipes->insert(pair<string, Pipe*>(pipeName, pipe));
    }

    ~NetworkModel()
    {

        map<string, Pipe *>::iterator itr;
        for (itr = pipes->begin(); itr != pipes->end(); itr++)
        {
            delete itr->second;
        }
        delete pipes;
        
    }
};

#endif